probe-rs/src/architecture/riscv/communication_interface.rs

```diff
@@ -0,0 +1,848 @@
+
+        let ebreak_cmd = 0b000000000001_00000_000_00000_1110011;
+
+        self.write_dm_register(Progbuf0(lw_command))?;
+        self.write_dm_register(Progbuf1(ebreak_cmd))?;
+
+        self.write_dm_register(Data0(address))?;
+
+        // Write s0, then execute program buffer
+        let mut command = AccessRegisterCommand(0);
+        command.set_cmd_type(0);
+        command.set_transfer(true);
+        command.set_write(true);
+
+        // registers are 32 bit, so we have size 2 here
+        command.set_aarsize(2);
+        command.set_postexec(true);
+
+        // register s0, ie. 0x1008
+        command.set_regno(0x1008);
+
+        self.write_dm_register(command)?;
+
+        let status: Abstractcs = self.read_dm_register()?;
+
+        if status.cmderr() != 0 {
+            return Err(RiscvError::AbstractCommand(
+                AbstractCommandErrorKind::parse(status.cmderr() as u8),
+            ));
+        }
+
+        // Read back s0
+        let value = self.abstract_cmd_register_read(0x1008)?;
+
+        self.abstract_cmd_register_write(0x1008, s0)?;
+
+        Ok(u32::from(value))
+    }
+
+    /// Perform memory write to a single location using the program buffer.
+    /// Only writes up to a width of 32 bits are currently supported.
+    fn perform_memory_write(
+        &mut self,
+        address: u32,
+        width: u8,
+        data: u32,
+    ) -> Result<(), RiscvError> {
+        // Backup registers s0 and s1
+        let s0 = self.abstract_cmd_register_read(0x1008)?;
+        let s1 = self.abstract_cmd_register_read(0x1009)?;
+
+        // assemble
+        //  lb s0, 0(s0)
+
+        //let o = 0; // offset = 0
+        //let b = 9; // base register -> s0
+        //let w = 0; // width
+        //let d = 9; // dest register -> s0
+        //let l = 0b11;
+
+        //let lw_command = bitpack!("oooooooooooobbbbb_www_ddddd_lllllll");
+        let mut sw_command: u32 = 0b0000000_01001_01000_000_00000_0100011;
+
+        // sw command -> sb s1, 0(s0)
+
+        // verify the width is supported
+        // 0 ==  8 bit
+        // 1 == 16 bit
+        // 2 == 32 bit
+
+        assert!(width < 3, "Width larger than 3 not supported yet");
+
+        sw_command |= (width as u32) << 12;
+
+        //if width == 2 {
+        //    sw_command = 0xc004;
+        //}
+
+        let ebreak_cmd = 0b000000000001_00000_000_00000_1110011;
+
+        self.write_dm_register(Progbuf0(sw_command))?;
+        self.write_dm_register(Progbuf1(ebreak_cmd))?;
+
+        // write value into s0
+        self.abstract_cmd_register_write(0x1008, address)?;
+
+        // write address into data 0
+        self.write_dm_register(Data0(data))?;
+
+        // Write s0, then execute program buffer
+        let mut command = AccessRegisterCommand(0);
+        command.set_cmd_type(0);
+        command.set_transfer(true);
+        command.set_write(true);
+
+        // registers are 32 bit, so we have size 2 here
+        command.set_aarsize(2);
```

Comment:
Consider creating enum / const for bus access types. Can be shared among multiple places.

E.g.

```rust
enum RiscvBusAccess {
    A8 = 0x0,
    A16  = 0x1,
    A32  = 0x2,
    A64  = 0x3,
    A128 = 0x4,
}
```

Reply:
Very good point, the interface is very consistent, so this can also be used e.g. for memory accesses.